# SDSS SQL Tutorial
---

### Names: [Insert names here]

**Before you do anything else, go to File -> Save a Copy in Drive. Make any requested edits to that copy.**

**New Code**

* Simple SQL query.
* Select objects
  * within a given range of RA and Dec.
  * based on object classification.
  * based on redshift.
  * based on numerical constraints (e.g., the color is above a certain value)
* Extract
  * RA, Dec
  * Magnitude at different bands
  * redshift
  * Object ID
* Return a sorted list of object. 
* Join tables. 
* Access the data returned by an SQL query within python.


In the first part of this lab, you were (briefly) introduced to the world of online astronomical databases, and shown how to perform simple searches. The funcionality of these databases extends well beyond what you saw in the previous lab. One of the powerful features of these databases is the ability to select out data that fit a certain set of criteria (beyond simply a location on the sky). This can include e.g., selecting all of the bright, well-resolved, nearby galaxies, in order to make your own Hubble classification diagram. Or selecting out all of the stars within a cluster (based on RA, Dec, and distance) in order to create an HR diagram of the cluster, and hence measure its age.

These complex queries are possible, although not with the simple query tools introduced earlier. To perform these tasks you need to learn about the Structured Query Language (SQL) a programming language for performing queries. To learn SQL, you will work through the SDSS SQL tutorial:
([http://skyserver.sdss.org/dr14/en/help/howto/search/searchhowtohome.aspx](http://skyserver.sdss.org/dr14/en/help/howto/search/searchhowtohome.aspx)). 

The tutorial can be completed entirely online, but SQL queries can also be completed within python. This notebook demonstrates how to call SQL queries within python. As you work through the online, record and execute your answers in the notebook below. 

### SQL Queries in Python

The next two cells show how to execute the initial SQL query in the tutorial

In [1]:
!pip install git+http://github.com/astropy/astroquery.git#egg=astroquery

  Cloning http://github.com/astropy/astroquery.git to /tmp/pip-install-zyp96t1b/astroquery_8d2fd4e1b82b437fb61777d6c1e6f022
  Running command git clone -q http://github.com/astropy/astroquery.git /tmp/pip-install-zyp96t1b/astroquery_8d2fd4e1b82b437fb61777d6c1e6f022
  Running command git submodule update --init --recursive -q
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 840 kB 5.7 MB/s 
     |████████████████████████████████| 54 kB 2.8 MB/s 
     |████████████████████████████████| 3.6 MB 46.1 MB/s 
  Created wheel for astroquery: filename=astroquery-0.4.6.dev7357-py3-none-any.whl size=4536622 sha256=a9eacfa2a24155db9433bdafa6f3c450756fb0cb30eb4b3367690fc884f62273
  Stored in directory: /tmp/pip-ephem-wheel-cache-sveiw22p/wheels/10/ed/6b/31374043d075d6321661bed3f565cacbdbc9d02179d2c3934e
  Created wheel for pyvo: filename=pyvo-1.2.1-py3-none-any.whl size=831658 sha256=

In [2]:
# First, suppress some warnings and import useful packages
import warnings
warnings.filterwarnings('ignore',module='astropy.io.votables.tree')
warnings.filterwarnings('ignore',message='.*unclosed..socket')

import astropy.units as u
from astropy.coordinates import SkyCoord, ICRS
from astroquery.sdss import SDSS #package that allows queries of the SDSS database


Basically, the SQL query is recorded in a string variable, and then this string is sent to SDSS using the `query_sql` function.

In [3]:
# First Query

# input the query as a string, and then submit the streing to SDSS
query_string = '''select ra,dec
from specObj
where ra BETWEEN 140 and 141 AND
dec BETWEEN 20 and 21'''

data = SDSS.query_sql(query_string,verbose=False)

# We can now print the results
print(data)

    ra       dec   
--------- ---------
140.00334 20.774178
140.00636 20.924277
140.01069 20.078161
140.01226 20.538016
140.01318 20.605621
140.01467 20.499697
140.01628 20.849852
140.01957 20.762055
140.02726 20.674341
 140.0331  20.25241
      ...       ...
140.95848 20.611813
 140.9595 20.832956
140.96916 20.290457
140.97144 20.045201
140.97225 20.121144
 140.9787 20.759427
140.97915 20.029572
140.98147 20.817946
140.99308 20.027302
140.99353 20.578518
140.99392 20.423977
Length = 328 rows


The result is an astropy Tables object [(http://docs.astropy.org/en/stable/api/astropy.table.Table.html#astropy.table.Table)](http://docs.astropy.org/en/stable/api/astropy.table.Table.html#astropy.table.Table), which is similar to a dictionary.

In [4]:
# The column names
print(data.keys())

# Access individual columns
print(data['ra'][:5])

['ra', 'dec']
    ra   
---------
140.00334
140.00636
140.01069
140.01226
140.01318


From here you can complete the rest of the tutorial, recording your answers in the cells below. The examples have also been copied here for your records. While tutorial covers a lot of material related to SQL, you only need to go as far as you can before the lab ends (you can save the rest for another time).

**Do I really want them to go as far as they can? Instead do I want them to reach a certain point??**

> **Practice 1:** What objects has the SDSS seen in a smaller area of the sky near ra = 140.5, dec = 20.5 (the same area you searched in the previous query)?
>
> Modify the previous query so it will return ra and dec of objects where the ra is between 140.25 and 140.75 and dec is between 20.25 and 20.75. How many objects did the query return?

In [5]:
# Practice 1
query_string = '''select ra,dec
from specObj
where ra between 140.25 and 140.75 
and dec between 20.25 and 20.75

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    ra       dec   
--------- ---------
140.25263 20.610694
140.26132  20.37615
140.26652  20.27486
140.26679 20.410276
140.26998 20.665789
140.29008 20.268252
140.29458 20.557766
140.29726 20.675356
140.30213 20.516802
140.30326 20.679935
      ...       ...
140.72306 20.266393
140.72367 20.351012
140.72577 20.342547
140.72754 20.583188
140.72894 20.652857
 140.7291 20.400367
140.73026 20.441962
140.74094 20.320367
140.74187  20.34313
140.74276 20.596196
140.74828 20.343699
Length = 89 rows


> **Practice 2:** Which of the objects you found in Practice 1 are galaxies? Modify your query so that it returns the ra, dec, and the best object ID for galaxies (and only galaxies) whose ra is between 194.25 and 194.75 and whose dec is between 2.25 and 2.75.

In [6]:
# Practice 2
query_string = '''select ra,dec
from specObj
where ra between 140.25 and 140.75 
and dec between 20.25 and 20.75
and class = "galaxy"

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    ra       dec   
--------- ---------
140.26652  20.27486
140.26679 20.410276
140.26998 20.665789
140.30213 20.516802
140.31007 20.374799
140.32603 20.627772
140.35157 20.735508
140.38724 20.602435
140.42225 20.443024
140.43593 20.450317
      ...       ...
140.70442 20.355897
140.71457 20.401091
140.72306 20.266393
140.72367 20.351012
140.72577 20.342547
140.72754 20.583188
 140.7291 20.400367
140.73026 20.441962
140.74094 20.320367
140.74187  20.34313
140.74828 20.343699
Length = 40 rows


In [7]:
# Sample Query: Logical Operators
query_string = '''select top 10
    z, ra, dec, bestObjID
from
    specObj
where
    class = 'galaxy'
    and z > 0.3
    and zWarning = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    z         ra       dec          bestObjID     
--------- --------- ---------- -------------------
0.3000001 184.86409  52.571314 1237657629529211079
0.3000009 222.98982  42.629408 1237658203986460871
0.3000009 146.39863  5.8334459 1237658423547920434
0.3000012 49.459411   0.847754 1237660241924063461
0.3000018 172.28104  10.234226 1237658493349789816
0.3000027 156.25024  7.6586271 1237658425162858683
0.3000027 231.07321  32.718373 1237662341082382595
0.3000033 197.37712  10.524518 1237662238545478566
0.3000035 184.90501 -3.3095095 1237650760782053596
0.3000037 242.52563  6.0693494 1237662198290186255


In [8]:
# Sample Query: Logical Operators
query_string = '''select top 10
    z, ra, dec, bestObjID
from 
    specObj
where
    (class = 'galaxy' or class = 'qso')
    and z > 0.3
    and zWarning = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    z         ra       dec          bestObjID     
--------- --------- ---------- -------------------
0.3000001 184.86409  52.571314 1237657629529211079
0.3000009 222.98982  42.629408 1237658203986460871
0.3000009 146.39863  5.8334459 1237658423547920434
0.3000012 49.459411   0.847754 1237660241924063461
0.3000018 172.28104  10.234226 1237658493349789816
0.3000027 156.25024  7.6586271 1237658425162858683
0.3000027 231.07321  32.718373 1237662341082382595
0.3000033 197.37712  10.524518 1237662238545478566
0.3000035 184.90501 -3.3095095 1237650760782053596
0.3000037 242.52563  6.0693494 1237662198290186255


In [9]:
# Sample Query: Logical Operators
query_string = '''select top 10
    ra, dec, modelMag_u, modelMag_g, modelMag_r, modelMag_i, modelMag_z, objID
from 
    photoObj
where
    type = 6
    and modelMag_u - modelMag_g < 0.5
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

       ra              dec        modelMag_u ... modelMag_z        objID       
---------------- ---------------- ---------- ... ---------- -------------------
21.7489625734203 20.8928509871811   24.47644 ...   20.93681 1237679541360656765
21.7146801359507 20.4432857186118    23.4531 ...   20.20991 1237679461900353900
22.3833537049599 20.0086760632325   23.31016 ...   21.33505 1237679504864968995
20.4532228010766 19.4146674875702   23.74691 ...    23.3264 1237679503790441055
20.4575029739604 19.4034512992811   22.09286 ...   21.46572 1237679503790441061
23.8528956918851  17.960344971995   25.89019 ...   23.56271 1237679459216982791
23.8678603354314 18.0722111769438     24.668 ...   23.14229 1237679459216982799
24.1930872176298 19.1999599866118    23.1454 ...   21.21239 1237679460827660758
 23.789412470134 18.2858004580844   24.12651 ...   20.76979 1237679502718141036
23.7877842920117 18.2587365072431   24.26909 ...   21.52649 1237679502718141037



> **Practice 3:** What are the reddest galaxies in the area of the sky near ra=141?
>
> Write a query to search for galaxies between ra = 140.9 and ra = 141.1 brighter than g=18.0 for which u-g>2.2. Retrieve the Object ID, ra, dec, and the five final magnitudes.

In [13]:
# Practice 3
query_string = '''select top 10
  objID, ra, dec, modelMag_u, modelMag_g, modelMag_r, modelMag_i, modelMag_z
from photoObj
where
  type = 3 and
  modelMag_g < 18.0 and
  modelMag_u-modelMag_g>2.2

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

       objID               ra        ... modelMag_i modelMag_z
------------------- ---------------- ... ---------- ----------
1237672132001530905 283.748738272469 ...   15.82844   15.29652
1237667206208422149 80.5427521003942 ...   15.95102   23.03563
1237667206208422152 80.5449438194449 ...   16.73747   19.53274
1237667206208421996 80.4552327793352 ...    12.0804   13.41948
1237667206208422131  80.444725770296 ...   15.65011   15.48841
1237672132001531061 283.867839890997 ...   10.87182   10.48374
1237672132001531103 283.868592770409 ...   15.70019    23.0845
1237654872683905062 154.849474331231 ...   15.18962   15.00237
1237655549668032618 242.972085394935 ...   9.720732   9.675129
1237652938327916898 313.683556336204 ...   15.75685   22.82109


> **Practice 4:** What are the highest-redshift quasars in the SDSS database?
>
> Write a query to search for quasars for which we have obtained spectra (search the specObj table) with redshifts greater than 4.5 and good measurements (zWarning=0). Retrieve each quasar's Photo ID, ra, dec, and redshift.

In [14]:
# Practice 4
query_string = '''select top 10
  ra, dec, z, bestobjID
from specobj
where
  z > 4.5 and
  zWarning = 0 and
  class = 'qso'

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

    ra        dec        z          bestobjID     
--------- ----------- -------- -------------------
226.87766   55.619792 4.500529 1237655465388999582
166.32833   19.832961  4.50241 1237668290162131387
167.10967  0.61852761 4.502814 1237674650996769779
145.02001   5.4419457 4.503839 1237674460417754011
116.39974   33.178191 4.504273 1237674289681072697
127.76246   52.592732 4.504608 1237651272960836116
 188.6143   29.449435 4.504811 1237667253480325869
34.927581 -0.19360219 4.505384 1237666407920829375
132.93219   23.535822 4.505914 1237665098459382395
 29.34542   2.0443026 4.507283 1237678618504397427


In [15]:
# Sample Query: Joining
query_string = '''select top 100
    x.plate, x.mjd,
    s.fiberID,
    p.modelMag_u, p.modelMag_g, p.ModelMag_r, p.ModelMag_i, p.ModelMag_z,
    p.ra, p.dec,
    s.z, p.ObjID
from photoObj p
join specObJ s on s.bestobjid = p.objid
join plateX x on x.plateID = s.plateID
where
    s.class = 'qso'
    and s.zwarning = 0
    and s.z between 0.3 and 0.4
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

plate  mjd  fiberID ...        dec            z            ObjID       
----- ----- ------- ... ----------------- --------- -------------------
  626 52057     503 ...  45.2064054294674 0.3999253 1237651251515883707
  616 52442     119 ...  54.0267704592088 0.3965549 1237651251510771972
  605 52353     494 ...  62.8356520577433 0.3862629 1237651251504087191
  626 52057     199 ...  43.8207993988105 0.3035774 1237651250979668146
  616 52442     470 ...  54.2451736767219 0.3662273 1237651250973573185
  624 52377     341 ...  47.5671491878967 0.3352541 1237651250440569147
  605 52353     126 ...  62.1738281193691 0.3297572 1237651250430345231
  605 52353     258 ...  62.3095704643739 0.3286313 1237651250430017615
  453 51915     220 ...  57.9527002259335 0.3511864 1237651249879908577
  453 51915     413 ...  60.2469556247499 0.3846239 1237651192436883690
  ...   ...     ... ...               ...       ...                 ...
  431 51877      51 ...  36.9454774209751 0.3028458 123765127241

> **Practice 5:** How can you look up image data, plates, and spectra of moderately bright galaxies?
>
> Write a query to find 100 galaxies for which we have spectra that have g magnitude between 17 and 17.4 and redshift less than 0.05. For each galaxy retrieve the object ID, the five magnitudes, the redshift, the plate/MJD number, and the fiber number.

In [16]:
# Practice 5
query_string = '''select top 100
  p.objid, p.modelmag_u, p.modelmag_g, p.modelmag_r, p.modelmag_i, p.modelmag_z, 
  s.z, s.fiberid,
  x.plateid
from photoobj p
join specobj s on s.bestobjid = p.objid
join platex x on x.plateid = s.plateid
where
  p.modelmag_g between 17 and 17.4 and
  s.z < 0.05 and
  s.class = 'galaxy'

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

       objid        modelmag_u modelmag_g ... fiberid       plateid      
------------------- ---------- ---------- ... ------- -------------------
1237658299534671894   18.54983   17.08479 ...     480 1125899951184832512
1237662302989648563    18.3209     17.367 ...     141 1899393197537257472
1237662500002726033   18.28371   17.28571 ...      62 1889260101278132224
1237658299534671969   18.19122   17.01253 ...     475 1125899951184832512
1237662500536385784   18.25008   17.37483 ...     627 1885882402111252480
1237658800965222553   18.50793   17.14219 ...     148 1148417950630307840
1237661435920187668   18.62061   17.24556 ...     602 1635932616383293440
1237661435920187571   18.87023   17.00496 ...     603 1635932616383293440
1237661434846380233   18.98217   17.12482 ...     583 1635932616383293440
1237661812810580159   18.59955   17.32438 ...     633 1813824802721392640
                ...        ...        ... ...     ...                 ...
1237661972268712183   18.43938   17.39

---

You can **stop here**, or you can continue if you have time.

---

In [ ]:
# Sample Query: Aggregate functions
query_string = ''' select
    min(dec) as min_dec, max(dec) as max_dec, avg(dec) as avg_dec
from
    photoObj
where
    run = 5112
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: The Group By command
query_string = '''select
    class, count(z) as num_redshift
from
    specObj
where
    z between 0.5 and 1
group by
    class
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: The Order By command
query_string = '''select mjd, plate
from
    plateX
where
    plate <= 275
order by mjd
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 7:** What are the northernmost and southernmost objects with spectra measured by the SDSS?

In [ ]:
# Practice 7
query_string = '''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 8:** What is the redshift of the nearest galaxy whose spectrum was measured by the SDSS with high confidence (zWarning=0)?
>
> Compare the distance you found to the Andromeda Galaxy (2 million light-years) and the Whirlpool Galaxy (37 million light years). Does the distance you found seem reasonable?

In [ ]:
# Practice 8
query_string = '''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 9:** What field has galaxies with the highest average redshifts in run=5112, camcol=1?
>
> Be sure you are searching fields (as run-camcol-field) for galaxies. Also look at how many spectrally measured galaxies are in the field - make sure you don't pick a field with only one or two galaxies! Also note that this query will probably take a long time to execute.

In [ ]:
# Practice 9
query_string = '''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

---

You can **stop here**, or continue if you have time.

### To turn in this lab, share the lab with me using the *Share* button in the upper right.

---

In [ ]:
# Sample Query: Functions
# Search for all objects within 5 arcminutes of ra=140, dec=20
query_string = '''select
    p.ObjID, p.ra, p.dec, p.u, p.g, p.r, p.i, p.z
from photoObj p
join dbo.fGetNearbyObjEq(140,20,5) n on n.objID = p.objID
where
    p.type = 3
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: Flags
# Find English names for flags of all stars around the point 175,1
query_string='''SELECT
    p.ObjID, p.flags, dbo.fPhotoFlagsN(p.flags)
FROM photoObj p
JOIN dbo.fGetNearbyObjEq(140,20,5) n ON n.objID = p.objID
WHERE
    p.type = 6
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

In [ ]:
# Sample Query: Flags
query_string='''SELECT
    p.ObjID, p.ra, p.dec, dbo.fPhotoFlagsN(p.flags)
FROM photoObj p
JOIN dbo.fGetNearbyObjEq(140,20,5) n ON n.objID = p.objID
WHERE
    (p.flags & dbo.fPhotoFlags('SATURATED')) = 0
'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 10:** In the field 5112-6-110, what percentage of all objects detected by SDSS are too close to the edge of their fields to be trusted?
>
> Hint: Use two searches, one with a flag and one without. Search run=5112, camcol=6, field = 119

In [ ]:
# Practice 10
query_string = '''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

> **Practice 11:** Choose a galaxy cluster from SkyServer's Famous Places tool. Write a query to select galaxies in the cluster, and only galaxies in the cluster.
>
> Hint: After you pick a cluster, use the Navigation Tool to examine the cluster. Guess which galaxies belong to the cluster - you should be able to tell just by looking. Click on 5-10 galaxies and save them in your online notebook. Open the notebook to look for features that the cluster galaxies have in common. Guess the center position and radius of the galaxies. Then, write a query that uses what you have learned to search for the cluster galaxies.

In [ ]:
# Practice 11
query_string = '''

'''

data = SDSS.query_sql(query_string,verbose=False)
print(data)

- Challenge 1: What percentage of galaxies have spectral redshifts measured? What percentage have photometric redshifts taken? What are the advantages of using photometric redshifts? Try to compare photometric and spectral redshifts; how accurate are photometric redshifts?
- Challenge 2: What are the limits in ra and dec of stripes 42 and 43, two of the SDSS's diagonal stripes?
- Challenge 3: Look at colors and spectra of stars, and find stars consistent with white dwarfs. Create a list of white dwarfs in the SDSS database.
- Challenge 4: What are the largest galaxies in the SDSS, in terms of size? Hint: Look at surface brightness and ellipticity.
- Challenge 5: Find all objects with spectra classified as unknown.
- Challenge 6: Find the broad absorption line (BAL) quasars in the SDSS database. At what redshift are most BAL quasars found?
- Challenge 7: Find variable stars in the SDSS (stars imaged more than once whose magnitude changed by more than 0.1 between observations). How variable are the stars?

SQL is a common language for querying databases. The International Virtual Observatory Association supports use of the Astronomy Database Query Language (ADQL), and is used by e.g. GAIA [https://www.gaia.ac.uk/data/gaia-data-release-1/adql-cookbook](https://www.gaia.ac.uk/data/gaia-data-release-1/adql-cookbook). Sincd ADQL is built on from SQL, much of the syntax is the same. 